File: customer-address-jpa/src/test/java/com/izertis/example/repository/jpa/CustomerRepositoryIntegrationTest.java
package com.izertis.example.repository.jpa;

import com.izertis.example.domain.Address;
import com.izertis.example.domain.Customer;
import com.izertis.example.domain.PaymentMethod;
import com.izertis.example.domain.PaymentMethodType;
import jakarta.persistence.EntityManager;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;

class CustomerRepositoryIntegrationTest extends BaseRepositoryIntegrationTest {

    @Autowired
    EntityManager entityManager;

    @Autowired
    CustomerRepository customerRepository;

    private Long id = 1L;

    @Test
    void findAllTest() {
        var results = customerRepository.findAll();
        Assertions.assertFalse(results.isEmpty());
    }

    @Test
    void findByIdTest() {
        var customer = customerRepository.findById(id).orElseThrow();
        Assertions.assertNotNull(customer.getId());
        Assertions.assertNotNull(customer.getVersion());
        Assertions.assertNotNull(customer.getCreatedBy());
        Assertions.assertNotNull(customer.getCreatedDate());
    }

    @Test
    void saveTest() {
        Customer customer = new Customer();
        customer.setName("Jane Smith");
        customer.setEmail("jane.smith@example.com");
        customer.setAddresses(List.of(
                        new Address()
                                .setStreet("456 Elm St")
                                .setCity("Othertown")
                )
        );

        // OneToMany paymentMethods owner: true
        var paymentMethods = new PaymentMethod();
        paymentMethods.setType(PaymentMethodType.VISA);
        paymentMethods.setCardNumber("6543210987654321");
        customer.addPaymentMethods(paymentMethods);

        // Persist aggregate root
        var created = customerRepository.save(customer);

        // reloading to get relationships persisted by id
        entityManager.flush();
        entityManager.refresh(created);
        Assertions.assertNotNull(created.getId());
        Assertions.assertNotNull(created.getVersion());
        Assertions.assertNotNull(created.getCreatedBy());
        Assertions.assertNotNull(created.getCreatedDate());

        Assertions.assertTrue(customer.getPaymentMethods().stream().allMatch(item -> item.getId() != null));
    }

    @Test
    void updateTest() {
        var customer = customerRepository.findById(id).orElseThrow();
        customer.setName("updated");
        customer.setEmail("updated@email.com");


        customer = customerRepository.save(customer);
        Assertions.assertEquals("updated", customer.getName());
        Assertions.assertEquals("updated@email.com", customer.getEmail());
    }

    @Test
    void deleteTest() {
        customerRepository.deleteById(id);
        var notFound = customerRepository.findById(id);
        Assertions.assertFalse(notFound.isPresent());
    }

}

File: kustomer-address-jpa/src/test/kotlin/com/izertis/example/repository/jpa/CustomerRepositoryIntegrationTest.kt
package com.izertis.example.repository.jpa

import com.izertis.example.domain.Address
import com.izertis.example.domain.Customer
import com.izertis.example.domain.PaymentMethod
import com.izertis.example.domain.PaymentMethodType
import jakarta.persistence.EntityManager
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired

class CustomerRepositoryIntegrationTest : BaseRepositoryIntegrationTest() {

    @Autowired
    lateinit var entityManager: EntityManager

    @Autowired
    lateinit var customerRepository: CustomerRepository

    private val id = 1L

    @Test
    fun findAllTest() {
        val results = customerRepository.findAll()
        assertFalse(results.isEmpty())
    }

    @Test
    fun findByIdTest() {
        val customer = customerRepository.findById(id).orElseThrow()
        assertNotNull(customer.id)
        assertNotNull(customer.version)
        assertNotNull(customer.createdBy)
        assertNotNull(customer.createdDate)
    }

    @Test
    fun saveTest() {
        val customer = Customer().apply {
            name = "Jane Smith"
            email = "jane.smith@example.com"
            addresses = mutableListOf(Address().apply {
                street = "456 Elm St"
                city = "Othertown"
            })
        }

        // OneToMany paymentMethods owner: true
        val paymentMethod = PaymentMethod().apply {
            type = PaymentMethodType.VISA
            cardNumber = "6543210987654321"
        }
        customer.addPaymentMethods(paymentMethod)

        // Persist aggregate root
        val created = customerRepository.save(customer)

        // reloading to get relationships persisted by id
        entityManager.flush()
        entityManager.refresh(created)

        assertNotNull(created.id)
        assertNotNull(created.version)
        assertNotNull(created.createdBy)
        assertNotNull(created.createdDate)

        assertTrue(customer.paymentMethods?.all { it.id != null } == true)
    }

    @Test
    fun updateTest() {
        val customer = customerRepository.findById(id).orElseThrow()
        customer.name = "updated"
        customer.email = "updated@email.com"

        val updated = customerRepository.save(customer)
        assertEquals("updated", updated.name)
        assertEquals("updated@email.com", updated.email)
    }

    @Test
    fun deleteTest() {
        customerRepository.deleteById(id)
        val notFound = customerRepository.findById(id)
        assertFalse(notFound.isPresent)
    }
}
